src/audio/volume/volume_hifi3.c

```diff
@@ -106,51 +125,68 @@ static void vol_s32_to_s24_s32(struct comp_dev *dev, struct audio_stream *sink,
 			       uint32_t frames)
 {
 	struct vol_data *cd = comp_get_drvdata(dev);
-	ae_f64 mult;
 	ae_f32x2 in_sample = AE_ZERO32();
 	ae_f32x2 out_sample;
 	ae_f32x2 volume;
-	size_t channel;
-	int shift = 0;
 	int i;
-	ae_int32 *in = (ae_int32 *)source->r_ptr;
-	ae_int32 *out = (ae_int32 *)sink->w_ptr;
-
-	/* Main processing loop */
-	for (i = 0; i < frames; i++) {
-		/* Processing per channel */
-		for (channel = 0; channel < sink->channels; channel++) {
-			/* Set source as circular buffer */
-			vol_setup_circular(source);
-
-			/* Load the input sample */
-			AE_L32_XC(in_sample, in, sizeof(ae_int32));
-
-			/* Load volume */
-			volume = (ae_f32x2)cd->volume[channel];
-
-			/* Multiply the input sample */
-			mult = AE_MULF32S_LL(volume, in_sample);
-
-			/* Multiplication of Q1.31 x Q1.31 gives Q1.63.
-			 * Now multiplication is Q8.16 x Q1.31, the result
-			 * is Q9.48. Need to shift right by one to get Q17.47
-			 * compatible format for round.
-			 */
-			out_sample = AE_ROUND32F48SSYM(AE_SRAI64(mult, 1));
-
-			/* Shift for S24_LE */
-			out_sample = AE_SRAA32RS(out_sample, shift);
-			out_sample = AE_SLAA32S(out_sample, shift);
-			out_sample = AE_SRAA32(out_sample, shift);
-
-			/* Set sink as circular buffer */
-			vol_setup_circular(sink);
-
-			/* Store the output sample */
-			AE_S32_L_XC(out_sample, out, sizeof(ae_int32));
-		}
+	ae_f64 mult0, mult1;
+	const int inc = sizeof(ae_f32x2);
+	const int channels_count = sink->channels;
+	const int samples = channels_count * frames;
+
+	/* store volume 4 times for xtensa 2-way intrinsic operations
+	 * to ensure the adsress is 8-byte aligned and avoid risk of
+	 * error loading of volume gain while the cd->vol would be set
+	 * as circular buffer
+	 */
+	for (i = 0; i < channels_count; i++) {
```

Comment:
this looks like common code that can be extracted into a function if it isn't too costly